### PR TITLE
Add initial EIP-1559 support

### PR DESCRIPTION
**Pull request by joshuajbouw**
Adds initial [EIP-1559](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md) support for the London hardfork.

There has been numerous changes done in this PR. Notably, I made a geth-like `NormalizedEthTransaction` which given the current fork will use that structure. This is important as for the first time instead of adding fields, 1 was taken away, and 2 additional were added so it led to a bit of a large redesign.

### CODE REVIEW

**Comment by joshuajbouw** [comment]
Unblocked as per https://github.com/aurora-is-near/aurora-engine/pull/280

File: engine-types/src/types.rs [hunk 1]
```diff
@@ -89,14 +90,20 @@ impl Wei {
     pub fn checked_add(self, other: Self) -> Option<Self> {
         self.0.checked_add(other.0).map(Self)
     }
+
+    pub fn checked_mul(self, other: Self) -> Option<Self> {
```
Comment (birchmd) [inline engine-types/src/types.rs:94]: Multiplying units of tokens (Wei in this case) does not make sense. The result would have units of `Money^2` which is meaningless (there is not a reasonable notion of "money area" as far as I know).

Comment (birchmd) [reply engine-types/src/types.rs]: Fixed in https://github.com/aurora-is-near/aurora-engine/pull/244/commits/08be2514183780684699a6fc27d8cbb25f58598e

File: engine/src/transaction/eip_1559.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
+use crate::prelude::{Vec, U256};
+use crate::transaction::eip_2930::AccessTuple;
+use aurora_engine_types::types::Wei;
+use ethabi::Address;
+use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
+
+/// Type indicator (per EIP-1559)
+pub const TYPE_BYTE: u8 = 0x01;
```
Comment (birchmd) [inline engine/src/transaction/eip_1559.rs:8]: Should be `0x02`?

File: engine/src/transaction/mod.rs [hunk 3]
```diff
@@ -1,107 +1,133 @@
...
-    }
+impl TryFrom<&[u8]> for EthTransactionKind {
+    type Error = ParseTransactionError;
 
-    pub fn gas_limit(&self) -> U256 {
-        match self {
-            Self::Legacy(tx) => tx.transaction.gas,
-            Self::AccessList(tx) => tx.transaction_data.gas_limit,
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        if bytes[0] == eip_2930::TYPE_BYTE {
```
Comment (birchmd) [inline engine/src/transaction/mod.rs:23]: Will need to add a case for `eip_1559::TYPE_BYTE`

File: engine/src/transaction/mod.rs [hunk 4]
```diff
@@ -1,107 +1,133 @@
...
-                    tx.access_list,
-                )
-            }
+impl From<EthTransactionKind> for NormalizedEthTransaction {
+    fn from(kind: EthTransactionKind) -> Self {
+        use EthTransactionKind::*;
+        match kind {
+            Legacy(tx) => Self {
+                address: tx.sender(),
+                chain_id: None, // This should be fine as it'll fail regardless.
```
Comment (birchmd) [inline engine/src/transaction/mod.rs:58]: This looks wrong. Why not properly derive the `chain_id` from the `v` field?

**Review by birchmd — CHANGES_REQUESTED** [review]

File: engine/src/transaction/eip_1559.rs (hunk 2 above)
Comment (joshuajbouw) [reply engine/src/transaction/eip_1559.rs:8]: Would need to double check, I do recall pulling that off of source.

File: engine/src/transaction/mod.rs (hunk 4 above)
Comment (joshuajbouw) [reply engine/src/transaction/mod.rs:58]: Yeah that should be updated.

**Pull request merged**